snmalloc-edp/src/rust-sgx-snmalloc-shim.cpp

```diff
@@ -0,0 +1,190 @@
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE
+
+#include <immintrin.h>
+#include <string.h>
+
+/***************************************************/
+/*** Imported symbols needed by snmalloc SGX PAL ***/
+/***************************************************/
+
+// from entry.S
+extern "C" size_t get_tcs_addr();
+
+// from Rust std
+extern "C" void __rust_print_err(const char *m, size_t s);
+extern "C" [[noreturn]] void __rust_abort();
+
+/*******************************************************/
+/*** Standard C functions needed by snmalloc SGX PAL ***/
+/*******************************************************/
+
+// definition needs to match GNU header
+extern "C" [[noreturn]] void abort() __THROW {
+    __rust_abort();
+}
+
+// definition needs to match GNU header
+extern "C" inline int * __attribute_const__ __errno_location (void) __THROW {
+    static int errno;
+    return &errno;
+}
+
+extern "C" {
+    static size_t HEAP_BASE;
+    static size_t HEAP_SIZE;
+};
+
+/***********************************/
+/*** snmalloc SGX PAL definition ***/
+/***********************************/
+
+#define SNMALLOC_PROVIDE_OWN_CONFIG
+#define SNMALLOC_SGX
+#define SNMALLOC_USE_SMALL_CHUNKS
+#define SNMALLOC_MEMORY_PROVIDER PALEdpSgx
+#define OPEN_ENCLAVE
+// needed for openenclave header:
+#define OE_OK 0
+
+#include "../snmalloc/src/snmalloc/pal/pal_noalloc.h"
+
+namespace snmalloc {
+void register_clean_up() {
+    // TODO: not sure what this is supposed to do
+    abort();
+}
+
+class EdpErrorHandler {
+  public:
+    static void print_stack_trace() {}
+
+    [[noreturn]] static void error(const char *const str) {
+        __rust_print_err(str, strlen(str));
+        abort();
+    }
+    static constexpr size_t address_bits = Aal::address_bits;
+    static constexpr size_t page_size = Aal::smallest_page_size;
+};
+
+using EdpBasePAL = PALNoAlloc<EdpErrorHandler>;
+
+class PALEdpSgx : public EdpBasePAL {
+   public:
+    const static size_t RAND_NUM_GEN_MAX_RETRIES = 64;
+    using ThreadIdentity = size_t;
+    static constexpr uint64_t pal_features = EdpBasePAL::pal_features | Entropy;
+
+    template <bool page_aligned = false>
+    static void zero(void *p, size_t size) noexcept {
+        memset(p, 0, size);
+    }
+
+    static inline uint64_t get_entropy64() {
+        long long unsigned int retry_count = 0;
+        long long unsigned int result = 0;
+        while (_rdrand64_step(&result) != 1 && retry_count < RAND_NUM_GEN_MAX_RETRIES) {
+            retry_count++;
+        }
+        return result;
```

Comment:
This is not correct, this will return something even if the last call to rdrand returned an error.

Reply:
@jethrogb I was thinking, maybe I can have an assertion at the end of the loop, something like the following:

long long unsigned int retry_count = 0;
long long unsigned int result = 0;
int status;
while ( retry_count < RAND_NUM_GEN_MAX_RETRIES && status = _rdrand64_step(&result) != 1) {
    retry_count++;
}
assert(retry_count < RAND_NUM_GEN_MAX_RETRIES);
return result;

How does the above look?

Reply:
```c
for (size_t retry_count = 0; retry_count < RAND_NUM_GEN_MAX_RETRIES; retry_count++) {
long long unsigned int result;
if (_rdrand64_step(&result) == 1) {
    return result;
}
}
EdpErrorHandler::error("no entropy available");
```